public/app/plugins/datasource/loki/querybuilder/components/LokiQueryEditorSelector.tsx

```diff
@@ -18,10 +19,24 @@ export const LokiQueryEditorSelector = React.memo<LokiQueryEditorProps>((props)
   const { query, onChange, onRunQuery, data } = props;
   const styles = useStyles2(getStyles);
   const [visualQuery, setVisualQuery] = useState<LokiVisualQuery>(query.visualQuery ?? getDefaultEmptyQuery());
+  const [parseModalOpen, setParseModalOpen] = useState(false);
+  const [pendingChange, setPendingChange] = useState<LokiQuery | undefined>(undefined);
 
   const onEditorModeChange = useCallback(
     (newMetricEditorMode: QueryEditorMode) => {
-      onChange({ ...query, editorMode: newMetricEditorMode });
+      const change = { ...query, editorMode: newMetricEditorMode };
+      if (newMetricEditorMode === QueryEditorMode.Builder) {
+        const result = buildVisualQueryFromString(query.expr);
```

Comment:
For some other PR probably, but I think this (parsing + error handling modal) could be moved into a separate hook and reused.